### PR TITLE
drivers: api: add i3c ; drivers: platform: stm32

### PR DESCRIPTION
**Pull request by gastmaier**
## Pull Request Description

Some new boards from STM32 add a new protocol called I3C.
This pr adds an I3C API layer and STM32 I3C platform implementation.
This resource from STM32 gives a good overview on the protocol:
https://www.st.com/resource/en/application_note/an5879-introduction-to-i3c-for-stm32-mcus-stmicroelectronics.pdf
The specification is available at:
https://www.mipi.org/mipi-i3c-hci-download

The API was tested on the following boards:
nucleo-h503rb (mainly) nucleo-h563zi
focusing on the transactions and IBI feature

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [X] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [X] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [X] I have performed a self-review of the changes
- [X] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [X] I have tested in hardware affected projects, at the relevant boards
- [X] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

File: include/no_os_i3c.h [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+	uint32_t ibi_payload_len;
+	/** I3C bus received IBI Payload. */
+	uint32_t ibi_payload;
+};
+
+/**
+ * @struct no_os_i3c_daa_lut_flags
+ */
+struct no_os_i3c_daa_lut_flags {
+	uint64_t addr		: 7;
```
[the remark this answers is not in the document]
Comment (gastmaier) [reply include/no_os_i3c.h:211]: You are right https://github.com/analogdevicesinc/no-OS/blob/f581385709844451d8443bc8867cc5fbf24deeaa/drivers/api/no_os_i3c.c#L220 is affected by endianess.
I will update to access via the .fields.

```
#include <stdint.h>
#include <stdbool.h>
#include <stdio.h>

struct no_os_i3c_init_param {
    struct no_os_i3c_bus_init_param     *bus;
    uint64_t    pid;
    uint8_t      addr;
    bool   is_static;
    bool   is_i3c;
};

struct no_os_i3c_daa_lut_flags {
    uint64_t addr           : 7;
    uint64_t is_attached    : 1;
    uint64_t pid        : 48;
};

union no_os_i3c_daa_lut {
    struct no_os_i3c_daa_lut_flags fields;
    uint64_t value;
};


int main () {
    struct no_os_i3c_init_param it = {
     .pid = 0x123456789ABC,
     .addr = 0x6E
    };

    union no_os_i3c_daa_lut daa;

    // This is problematic
    daa.value = (it.pid << 48) | it.addr;
    // This is not
    //daa.fields.pid = it.pid;
    //daa.fields.addr = it.addr;

    printf("pid %llx ; addr %x ; raw %llx\n", daa.fields.pid, daa.fields.addr, daa.value);

    return 0;
}
```

```
mips-linux-gnu-gcc -static -o main-mips main.c
powerpc-linux-gnu-gcc -static -o main-power main.c
gcc -static -o main main.c

qemu-ppc ./main-power
qemu-mips ./main-mips
./main
```

**Comment by gastmaier** [comment]
V0 -> V1

* Fix bitfields endianess bug by dropping the the union and setting the struct fields directly.
* Fix memset bug lacking struct size
* Encapsulating variables macros around parenthesis.
* Astyle auto fixes

**Comment by gastmaier** [comment]
V1 -> V2

* Doxygen warnings fixes.

Compare with d41ad8eb4845e329c1640a605ae277f5036b30b6..79dede6d97113f3bc8c93e0687490380f6cf5a53

File: include/no_os_i3c.h [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+	uint32_t ibi_payload;
+};
+
+/**
+ * @struct no_os_i3c_daa_lut
+ * @brief Stores the PID + DA information to look-up during the DAA.
+ * Should be used only for device initialization, since it may
+ * go outdated.
+ */
+struct no_os_i3c_daa_lut {
```
Comment (buha) [inline include/no_os_i3c.h:213]: I still see this one containing bit fields and i think they should also be remove

Besides the endianess issue there is also a risk when serializing this info and sending it to another device (like a PC) that runs software to interpret the info. The PC software could be compiled with different compiler than the firmware and the packing, order, padding etc. of the structures with bitfields may differ.

File: include/no_os_i3c.h [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+union no_os_i3c_bcr {
+	struct no_os_i3c_bcr_flags fields;
+	uint8_t value;
+};
+
+/**
+ * @struct no_os_i3c_irq_flags
+ */
+struct no_os_i3c_irq_flags {
+	uint8_t IBI	: 1;
```
Comment (buha) [inline include/no_os_i3c.h:181]: and here.

File: include/no_os_i3c.h [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+	NO_OS_I3C_ADDR_SLOT_I3C_DEV,
+	/** Address slot mask */
+	NO_OS_I3C_ADDR_SLOT_STATUS_MASK = 3,
+};
+
+/**
+ * @struct no_os_i3c_bcr_flags
+ */
+struct no_os_i3c_bcr_flags {
+	uint8_t MAX_DATA_SPEED_LIMITATION	: 1;
```
Comment (buha) [inline include/no_os_i3c.h:159]: and here.

**Comment by gastmaier** [comment]
V2 -> V3 

* Completely remove from struct bitfields
* Minor comment fomatting

Tested on hw
git diff 79dede6d97113f3bc8c93e0687490380f6cf5a53..14eba02a396a44ca3d7a1284fb930e4677d4c55f

**Review by buha — APPROVED** [review]

**Pull request merged**